File: libs/store-manager/src/main/kotlin/mono/store/manager/StoreManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.store.manager

import kotlinx.browser.localStorage
import kotlinx.browser.window
import mono.store.manager.migrations.MigrateTo2
import mono.store.manager.migrations.Migration
import org.w3c.dom.StorageEvent
import org.w3c.dom.get
import org.w3c.dom.set

/**
 * A class for managing storage.
 */
internal class StoreManager private constructor() {
    private val keyToObserverMap: MutableMap<String, StoreObserver> = mutableMapOf()
    private val migrations: Map<Int, Migration> =
        listOf<Migration>(
            MigrateTo2
        ).associateBy { it.targetVersion }

    init {
        migrate()

        window.onstorage = ::onStorageChange
    }

    internal fun getKeys(predicate: (String) -> Boolean): Sequence<String> = sequence {
        for (index in 0 until localStorage.length) {
            val key = localStorage.key(index)
            if (key != null && predicate(key)) {
                yield(key)
            }
        }
    }

    private fun migrate() {
        // handle migration, default does not have app version
        val currentDbVersion = get(StoreKeys.DB_VERSION)?.toIntOrNull() ?: 1

        for (version in currentDbVersion + 1..DB_VERSION) {
            migrations[version]?.migrate(this)
        }

        set(StoreKeys.DB_VERSION, DB_VERSION.toString())
    }

    fun set(key: String, json: String) {
        localStorage[key] = json
    }

    fun get(key: String): String? = localStorage[key]

    fun remove(key: String) {
        localStorage.removeItem(key)
    }

    fun setObserver(key: String, observer: StoreObserver) {
        if (key in keyToObserverMap) {
            error("$key's observer is set!")
        }
        keyToObserverMap[key] = observer
    }

    fun removeObserver(key: String) {
        keyToObserverMap.remove(key)
    }

    private fun onStorageChange(event: StorageEvent) {
        val key = event.key ?: return
        keyToObserverMap[key]?.onChange(key, event.oldValue, event.newValue)
    }

    companion object {
        private const val DB_VERSION = 2

        private var instance: StoreManager? = null

        fun getInstance(): StoreManager {
            val nonNullInstance = instance ?: StoreManager()
            instance = nonNullInstance
            return nonNullInstance
        }
    }
}

File: monosketch-svelte/src/lib/mono/store-manager/store-manager.ts
import { StoreKeys, DB_VERSION } from './consts';
import type { Migration } from './migrations/interface';
import { MigrateTo2 } from './migrations/migrate2';

/**
 * An interface for observing storage change.
 */
export interface StoreObserver {
    onChange(key: string, oldValue: string | null, newValue: string | null): void;
}

/**
 * A class for managing storage.
 */
export class StoreManager {
    private static instance: StoreManager | null = null;

    public static getInstance(): StoreManager {
        if (!StoreManager.instance) {
            StoreManager.instance = new StoreManager();
        }
        return StoreManager.instance;
    }

    private keyToObserverMap: { [key: string]: StoreObserver } = {};

    private migrations: { [targetVersion: number]: Migration } = {
        2: MigrateTo2,
    };

    private constructor() {
        this.migrate();

        window.onstorage = this.onStorageChange;
    }

    public getKeys(predicate: (key: string) => boolean): string[] {
        const keys: string[] = [];
        for (let index = 0; index < localStorage.length; index++) {
            const key = localStorage.key(index);
            if (key && predicate(key)) {
                keys.push(key);
            }
        }
        return keys;
    }

    private migrate(): void {
        const currentDbVersion = parseInt(this.get(StoreKeys.DB_VERSION) ?? '1');
        for (let version = currentDbVersion + 1; version <= DB_VERSION; version++) {
            const migration = this.migrations[version];
            if (migration) {
                migration.migrate(this);
            }
        }
        this.set(StoreKeys.DB_VERSION, DB_VERSION.toString());
    }

    public set(key: string, json: string): void {
        localStorage.setItem(key, json);
    }

    public get(key: string): string | null {
        return localStorage.getItem(key);
    }

    public remove(key: string): void {
        localStorage.removeItem(key);
    }

    public setObserver(key: string, observer: StoreObserver): void {
        if (this.keyToObserverMap[key]) {
            throw new Error(`${key}'s observer is already set!`);
        }
        this.keyToObserverMap[key] = observer;
    }

    public removeObserver(key: string): void {
        delete this.keyToObserverMap[key];
    }

    private onStorageChange = (event: StorageEvent) => {
        const key = event.key;
        if (key && this.keyToObserverMap[key]) {
            this.keyToObserverMap[key].onChange(key, event.oldValue, event.newValue);
        }
    };
}

/**
 * A storage document for a specific path.
 * This helps the storage have a structure similar to a directory in the file system.
 */
export class StorageDocument {
    public static get(path: string): StorageDocument {
        return (
            StorageDocument.documents[path] ??
            (StorageDocument.documents[path] = new StorageDocument(
                path,
                StoreManager.getInstance(),
            ))
        );
    }

    private static documents: { [path: string]: StorageDocument } = {};

    private constructor(
        private readonly path: string,
        private storeManager: StoreManager,
    ) {}

    private getFullPath(key: string): string {
        return StoreKeys.getPath(this.path, key);
    }

    public childDocument(key: string): StorageDocument {
        return StorageDocument.get(this.getFullPath(key));
    }

    public get(key: string, defaultValue: string | null = null): string | null {
        return this.storeManager.get(this.getFullPath(key)) ?? defaultValue;
    }

    public set(key: string, value: string): void {
        this.storeManager.set(this.getFullPath(key), value);
    }

    public remove(key: string): void {
        this.storeManager.remove(this.getFullPath(key));
    }

    public setObserver(key: string, observer: StoreObserver): void {
        this.storeManager.setObserver(this.getFullPath(key), observer);
    }

    public removeObserver(key: string): void {
        this.storeManager.removeObserver(this.getFullPath(key));
    }

    public getKeys(predicate: (key: string) => boolean): string[] {
        return this.storeManager.getKeys(predicate);
    }
}
